File: mycollab-web/src/main/java/com/esofthead/mycollab/vaadin/mvp/UrlResolver.java
package com.esofthead.mycollab.vaadin.mvp;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.esofthead.mycollab.core.MyCollabException;
import com.esofthead.mycollab.core.utils.BeanUtility;
import com.esofthead.mycollab.web.AppContext;
import com.vaadin.ui.Window;

public class UrlResolver {
	private static Logger log = LoggerFactory.getLogger(UrlResolver.class);

	private Map<String, UrlResolver> subResolvers;

	public void addSubResolver(String key, UrlResolver subResolver) {
		if (subResolvers == null) {
			subResolvers = new HashMap<String, UrlResolver>();
		}
		subResolvers.put(key, subResolver);
	}

	public void handle(String... params) {
		try {
			if (params != null && params.length > 0) {
				String key = params[0];
				if (subResolvers == null) {
					handlePage(params);
				} else {
					UrlResolver urlResolver = subResolvers.get(key);
					if (urlResolver == null) {
						throw new MyCollabException(
								"Can not register resolver key " + key
										+ " for Resolver: " + this);
					} else {
						preHandle();
						List<String> paramList = Arrays.asList(params).subList(
								1, params.length);
						log.debug("Handle url in resolver: " + urlResolver);
						urlResolver.handle(paramList.toArray(new String[0]));
					}
				}
			} else {
				handlePage(new String[0]);
			}
		} catch (Exception e) {
			log.error("Error while navigation", e);
			AppContext
					.getApplication()
					.getMainWindow()
					.showNotification("Information",
							"The record is not existed",
							Window.Notification.TYPE_HUMANIZED_MESSAGE);
		}
	}

	protected void preHandle() {
	}

	protected void handlePage(String... params) {
		log.debug("Handle page: " + this + " with params: "
				+ BeanUtility.printBeanObj(params));
	}
}

File: mycollab-ui/src/main/scala/com/esofthead/mycollab/vaadin/mvp/UrlResolver.scala
package com.esofthead.mycollab.vaadin.mvp

import com.esofthead.mycollab.core.MyCollabException
import com.esofthead.mycollab.vaadin.ui.NotificationUtil
import org.slf4j.{Logger, LoggerFactory}

import scala.annotation.varargs
import scala.collection.mutable._

/**
  * @author MyCollab Ltd
  * @since 5.0.9
  */
abstract class UrlResolver {
  private var subResolvers: Map[String, UrlResolver] = _
  protected var defaultUrlResolver: UrlResolver = _

  def addSubResolver(key: String, subResolver: UrlResolver) {
    if (subResolvers == null) {
      subResolvers = Map[String, UrlResolver]()
    }
    subResolvers += key -> subResolver
  }

  def getSubResolver(key: String): UrlResolver = subResolvers(key)

  @varargs def handle(params: String*): Unit = {
    try {
      if (params.length > 0) {
        val key = params(0)
        if (subResolvers == null) {
          handlePage(params: _*)
        }
        else {
          val urlResolver = subResolvers.get(key)
          urlResolver match {
            case Some(value) => value.handle(params.tail: _*)
            case None => {
              if (defaultUrlResolver != null) {
                defaultUrlResolver.handle(params: _*)
              }
              else {
                throw new MyCollabException(String.format("Can not register resolver key %s for Resolver: %s", key, this))
              }
            }
          }
        }
      }
      else {
        handlePage()
      }
    }
    catch {
      case e: Exception => {
        UrlResolver.LOG.error("Error while navigation", e)
        defaultPageErrorHandler()
        NotificationUtil.showRecordNotExistNotification()
      }
    }
  }


  protected def defaultPageErrorHandler(): Unit

  /**
    * @param params
    */
  @varargs protected def handlePage(params: String*): Unit = {}
}

object UrlResolver {
  val LOG: Logger = LoggerFactory.getLogger(classOf[UrlResolver]);
}
